### DISCUSSION TITLE
Create Server Action Closure on the Server

### DISCUSSION THREAD

**Discussion by jeengbe** [vercel/next.js#59911] [App Router]
It would be nice to have sort of "Server Action factory functions" in that they take data and return a Server Action that closes over that data. This way, internal data required for certain actions (such as user ID, for instance) would never be sent to the client.

---

```ts
// action.ts
'use server';

export default (userId: string) => (formData: FormData) => {
  // Do stuff...
}
```

```tsx
// server-component.tsx
import action from './action';

export default function Page({ userId }: { userId: string }) {
  return <form action={action(userId)}> ... </form>;
}
```

---

I see several drawbacks and technical challenges to the trivial solution that might render this idea impossible:

- The server becomes stateful: It has to keep track of dynamically created functions (the Actions). 
- Load balancing becomes difficult: Due to the above listed point, the server instance that served the page must also handle any bound Server Actions since only that node knows the value of a variable that an Action closes over.
- Lifetime: How long should the Action and values it closes over be kept in memory? If a user executes a Server Action after it has been cleaned up end up with an error.
- Problems with SSR?

The benefit of this is that Actions would more ergonomic in that you don't have to serialize all required information and (crucial!) send it to the client.

---

If this is not possible, an alternative would be to support currying within server functions, so that the example found here: https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#passing-additional-arguments could be replaced with:

```tsx
'use client'
 
import { updateUser } from './actions'
 
export function UserProfile({ userId }: { userId: string }) {
  const updateUserWithId = updateUser(userId)
 
  return (
    <form action={updateUserWithId}>
      <input type="text" name="name" />
      <button type="submit">Update User Name</button>
    </form>
  )
}
```

```ts
'use server'
 
export const updateUser = (userId) => async (formData) {
  // ...
}
```

which is almost the same, minus that the data exclusively stays on the server, but the function becomes easier to use.

**Comment by benevbright** [comment]
currying doesn't work for me too.

**Comment by hjumeau** [comment]
I also tried to use currying in a server action without success because server actions must expose asynchronous functions. Currying in these cases is more natural and simple for me.
